service/storage_service.cc

```diff
@@ -1181,116 +1171,169 @@ class topology_coordinator {
         co_return std::move(guard);
     }
 
-    future<> set_tablet_transition_stage(std::vector<canonical_mutation>& out, group0_guard& guard,
-                                         locator::tablet_transition_stage stage) {
-        auto tm = get_token_metadata_ptr();
-        for (auto&& [table, tmap] : tm->tablets().all_tables()) {
-            co_await coroutine::maybe_yield();
-            auto s = _db.find_schema(table);
-            for (auto&& [tablet, trinfo] : tmap.transitions()) {
-                co_await coroutine::maybe_yield();
-                auto last_token = tmap.get_last_token(tablet);
-                out.emplace_back(
-                    replica::tablet_mutation_builder(guard.write_timestamp(), s->ks_name(), table)
-                        .set_stage(last_token, stage)
-                        .build());
-            }
+    // Represents a two-state state machine which changes monotonically
+    // from "not executed" to "executed successfully". This state
+    // machine is transient, lives only on this coordinator.
+    // The transition is achieved by execution of an idempotent async
+    // operation which is tracked by a future. Even though the async
+    // action is idempotent, it is costly, so we want to avoid
+    // re-executing it if it was already started by this coordinator,
+    // that's why we track it.
+    using background_action_holder = std::optional<future<>>;
+
+    // Transient state of tablet migration which lives on this coordinator.
+    // It is guaranteed to die when migration is finished.
+    // Next migration of the same tablet is guaranteed to use a different instance.
+    struct tablet_migration_state {
+        background_action_holder streaming;
+    };
+
+    std::unordered_map<locator::global_tablet_id, tablet_migration_state> _tablets;
+
+    // Set to true when any action started on behalf of a background_action_holder
+    // for any tablet finishes, or fails and needs to be restarted.
+    bool _tablets_ready = false;
+
+    seastar::gate _async_gate;
+
+    // This function drives background_action_holder towards "executed successfully"
+    // by starting the action if it is not already running or if the previous instance
+    // of the action failed. If the action is already running, it does nothing.
+    // Returns true iff background_action_holder reached the "executed successfully" state.
+    bool advance_in_background(locator::global_tablet_id gid, background_action_holder& holder, const char* name,
+                               std::function<future<>()> action) {
+        if (!holder || holder->failed()) {
+            holder = futurize_invoke(action)
+                        .finally([this, g = _async_gate.hold(), gid, name] () noexcept {
+                slogger.trace("raft topology: {} for tablet {} resolved.", name, gid);
+                _tablets_ready = true;
```

Comment:
e.g. _tablets_ready.insert(gid).

If you don't want to handle failures here, you can pre-allocate the node to insert and just merge it here, using std::unordered_set::merge.

Yet another option is to use bitsets with summaries.

Of course, as an optimization, it can be deferred.

Reply:
I actually started with a set, but it turned out we I don't need it because scheduling is done in the load balancer, which is stateless. If it was stateful, then it would make sense to notify it about changed. 

The nice thing about load balancer being stateless is that the code is very simple and easy to reason about.

Reply:
It won't work when we manage 100k tablets.